### ISSUE TITLE
Potential grammar change to review

### ISSUE THREAD

**Issue by bd82** [jhipster/jhipster-core#241]
In 3.02 many of the application config values were changed to be alpha numeric
instead of alphabetic.

See: https://github.com/jhipster/jhipster-core/commit/05a907747870a9d2d7a662c773752419d475e1e0

The Chevrotain implementation allows alpha numeric values
with the **single constraint** that they must start with a letter.

e.g:
```
application foo {
    config {
       // valid (at least syntax) in 3.02 
       databaseType 666
    }
}

application foo {
    config {
       // valid (at least syntax) in 3.02 
       databaseType 111mssql
    }
}
```

With the Chevrotain Parser the two examples above are invalid.
I am guessing that these examples (starting with number / number only) should be invalid.
Please review and let me if these keys should have values starting with numbers.

**Comment by bd82** [comment]
@MathieuAA please have a look.

**Comment by ruddell** [comment]
The only JHipster variables I'm aware of that have numbers are `oauth2` (`authenticationType`) and `h2Disk`/`h2Memory` (`devDatabaseType`).

Regarding variables set by the user, `baseName` and `uaaBaseName` can start with letters or numbers

**Comment by MathieuAA** [comment]
@ruddell @bd82 They may be the only examples where there are letters & numbers, but we're not safe from tools having a number in their name... The safe option would be to support having them in the option names.
However, the chance of having a number first is pretty low, so we can have a letter first then (almost) whatever next.

**Comment by bd82** [comment]
Well since having a number first was only supported in the most recent patch version
and the chance of valid values starting with numbers is small I suggest we will handle this
as a **patch** if and when this issue will be raised by users.

I can solve this mini ambiguity (numbers vs identifiers starting with number) but I'm not sure its worth dealing with it right now if it is mostly theoretical.

**Comment by bd82** [comment]
If this approach is acceptable we can close this issue.

**Comment by MathieuAA** [comment]
Well for me it 50/50... Let's fix it once and for all for the two properties so that we don't have to deal with this later.

**Comment by bd82** [comment]
I will provide the "fix" in a PR
and you can decide if to merge it or not.

I'm just not fond of adding ambiguities (even simple resolvable ones)
When I'm not sure there is a real need.

**Comment by bd82** [comment]
I just had another look at the two properties @ruddell mentioned: 
* baseName 
* uaaBaseName

**Comment by MathieuAA** [comment]
This is tricky. Technically, they can be used as paths (and they are used as such), but let's not complicate things. Let's not add up more things... There should be rules for these two properties., someday.

**Comment by ruddell** [comment]
`baseName` can only contain letters, numbers and `_` , it is [validated here in the prompts](https://github.com/jhipster/generator-jhipster/blob/master/generators/generator-base.js#L2151-L2153)

**Comment by bd82** [comment]
o.k. hopefully tomorrow I will find a bit of time to try to implement a baseName property
that matches `/^([a-zA-Z0-9_]+)$/` instead of `/^([a-zA-Z][a-zA-Z0-9_]*)$/`

**Comment by bd82** [comment]
btw @ruddell that validation would also match an empty string:
```
> /^([a-zA-Z0-9_]*)$/.test('')
true
```

is that intentional?

**Comment by ruddell** [comment]
No it should not be blank.  The prompts have a default value of the current directory name so users can't input a blank value

**Comment by bd82** [comment]
Hello again.

I've implemented the logic required to allow identifiers starting with digits as the value for
the baseName property. #245

I am not pleased with the added complexity it contains and it also makes some assumptions
that could make the grammar more fragile.
There may be other approaches to this but I cannot recommend that this approach would be merged.

### Recommendation

Support **both** identifiers and string literals as the values for the baseName property.
* 99.9% of use cases would use the regular Identifiers which start with a letter.
   `baseName foo2`
* The 0.1% that need special characters can put it in a string literal.
  `baseName "123foo2"`

This won't require any grammar change, only to update the config validations logic
to allow both String Literals and Identifiers in that point in the grammar.

**Comment by MathieuAA** [comment]
@bd82 to sum up, you suggest not supporting having base names beginning with digits due to the implementation complexity?
About your recommendation, how do you propose this should be done?
In `configPropsValidations`, `baseName` is checked according to a regex.
Is your idea about using the `this.OR` method from the Parser but using it locally if you define a `baseNameDeclaration` rule? This way you should have more control over the type of each of your properties (the cost being having to define all your property rules)...

I feel there are some hard limitations from chevrotain when defining a grammar ([here](https://github.com/jhipster/jhipster-core/pull/245/files#diff-166390481de618205324cc3c0b12fb6bR387) for instance... gotta study that).

**Comment by bd82** [comment]
>  About your recommendation, how do you propose this should be done?

Syntax remains the same, configuration values can already be a string literal.
We just need the validations to allow both cases of string literals **and** of identifier.

> In configPropsValidations, baseName is checked according to a regex.
Is your idea about using the this.OR method from the Parser but using it locally if you define a baseNameDeclaration rule?

No, I only want to change the validations, not expand the config keys types into the grammar like in the pegjs grammar.

> I feel there are some hard limitations from chevrotain when defining a grammar (here for instance... gotta study that).

The limitation is the same high level one of having a **separate** lexing phase, so it is generally true
for any Parsers of this class, Each Token must be identified **in advance** without having any context information from the Parser.

Maybe there is a better way to do so, @colameo how did you implement this in the Editor?
The framework you are using has a separate Lexing phase correct?

I will provide an example how to implement my recommendation.

**Comment by bd82** [comment]
> Maybe there is a better way to do so, @colameo how did you implement this in the Editor?
The framework you are using has a separate Lexing phase correct?

I see a baseName property starting with a digit is not supported (yet) in the IDE
https://github.com/jhipster/jhipster-ide/blob/master/jdl/io.github.jhipster.jdl/src/io/github/jhipster/jdl/JDL.xtext#L344

**Comment by colameo** [comment]
yes, it's not supported because I thought identifiers must be at least valid "java identifiers" or am i wrong? However, It's difficult (at least for me) to implement features without a clear spec... (and an EBNF is not enough!), I would prefer just a plain text documentation with a structure, possible inputs, optionality (mandator/optional definitions), a short description etc. (like we had once) and some examples.

**Comment by colameo** [comment]
@bd82 I suppose the way how it's implemented in the IDE is quite different from yours. The grammar in the IDE is less restrictive as I've implemented validations to check the validity of parameter values. I also had to implement "special logic" in the content assist to make things work nice for users.

**Comment by bd82** [comment]
@colameo 

> yes, it's not supported because I thought identifiers must be at least valid "java identifiers" or am i wrong?

I thought so too but apparently the baseName property's value may start with a digit.

> it's difficult (at least for me) to implement features without a clear spec... 

I feel your pain, I'm experimenting with building a Parser for Less (css preprocessor)
and in that case there is not even an EBNF 😄.

But I also understand how much work it is to define a high quality spec and keep it updated
so that is often not possible in smaller projects as opposed to something like ECMA-262 standard committee (JavaScript).

>  I suppose the way how it's implemented in the IDE is quite different from yours. The grammar in the IDE is less restrictive as I've implemented validations to check the validity of parameter values

I actually did it the same as you, known possible key names and generic value options which are examined in a post processing phase.

**Comment by bd82** [comment]
@colameo given that baseName property may start with a digit.
* How would you disambiguate it from an integer?
*Aafaik XText uses Antlr Parser generator, so there is a separate lexing phase which would cause such an ambiguity.

Maybe you have a better idea than my overly complicated one shown [here](https://github.com/jhipster/jhipster-core/pull/245/files#diff-166390481de618205324cc3c0b12fb6bR66)?
* Basically I only allowed parsing an Identifier starting with a number
immediately after encountering a "baseName" token.

My thought that this is too fragile and we should allow **both** identifiers and string literals
as valid types for the "baseName" property. and  have the customer use string literals to wrap any names that don't fit a standard identifier.

**Comment by MathieuAA** [comment]
@colameo @bd82 Another option would be not to respect what's actually permitted by the generator and enforce some things. If this is such a pain in the ass, we can just not support base names beginning with digits... The only thing that's actually required is to document that (there's a PR for that documenting every option for applications, the same way there are annexes for field types).

**Comment by deepu105** [comment]
@MathieuAA I agree if it needs too many workarounds to achieve some edge cases then we can prohibit that in JDL and even in JHipster if needed

**Comment by bd82** [comment]
I think we can provide this functionality using the string literal workaround.
But if its an edge case that we are not sure is even needed maybe deal with a workaround 
once someone actually requests the capability...

**Comment by colameo** [comment]
...to allow string and literal for the same property would be (from my point of view) even worst.
however, I could manage to handle that case:
<img width="421" alt="screen shot 2018-07-05 at 21 35 22" src="https://user-images.githubusercontent.com/629760/42344110-996a06a4-809b-11e8-8bb3-15acf383f4b6.png">

**Comment by bd82** [comment]
> I could manage to handle the case:

Can you share the "how?" perhaps I can apply your solution and it would be less messy than mine...

**Comment by colameo** [comment]
...
JdlApplicationParameterValue:
	('['? identifiers+=JdlApplicationParameterLiteral (JdlParamValueSeparator identifiers+=JdlApplicationParameterLiteral)* ']'?) 
		| stringValue=STRING | numberValue=INT | version=JdlApplicationParameterVersion;

JdlApplicationParameterLiteral:
	ID | NUMID | JdlReserverdWords;
...
terminal ID: ('a'..'z'|'A'..'Z'|'_'|'-') ('a'..'z'|'A'..'Z'|'_'|'-'|'0'..'9')*;
terminal NUMID returns ecore::EString: INT? ID;
--------------------------
...everything else is handled with a validation.

But hey, if everybody agrees, lets change that in a string literal and handle the input with validations.

**Comment by bd82** [comment]
Thanks @colameo. I thought of implementing it as you suggested.
but while its not complex I am afraid it is still fragile
because an INT followed by a an Identifier must now have a whitespace separating the two.
And that requirement is for everywhere in the grammar.

I tried to limit the places where these special NUMIDs could even be lexed to reduce the fragility
But ended up increasing the complexity... 😢 

Maybe I'm (again) too used to programing languages syntax where identifiers cannot start
with numbers and it is not too bad in the case of a simple config language like JDL.

I think we should **defer** investing any more time in this issue until we encounter
a real use case where someone actually uses:

```
application {
  config {
    baseName 123foo
  }
}
```

And if/when that occurs solve it in the **most simple** manner by **also** allowing stringLiterals
```
application {
  config {
    baseName 123foo // not valid ID
  }
}

application {
  config {
    baseName "123foo" // valid using StringLiteral as a workaround, ugly for 0.1% of use cases.
  }
}

application {
  config {
    baseName toto // Valid regular ID suitable for 99.9% of cases.
  }
}


```

There is no cause to add complexities and irregularities

**Comment by colameo** [comment]
...just to make things clear, spaces would not be allowed in my parse rule (Chevrotain seems to be different here):

<img width="451" alt="screen shot 2018-07-07 at 15 44 36" src="https://user-images.githubusercontent.com/629760/42411483-c932972c-81fc-11e8-8983-5193b207616b.png">

However, I can make both cases (string and identifier literals) valid inputs:

<img width="424" alt="screen shot 2018-07-07 at 15 46 01" src="https://user-images.githubusercontent.com/629760/42411490-e17a0f36-81fc-11e8-8ffc-abe74fb6a210.png">
<img width="448" alt="screen shot 2018-07-07 at 15 44 25" src="https://user-images.githubusercontent.com/629760/42411501-0e2691b2-81fd-11e8-8e72-dd992637eed9.png">

Of course, if we would not allow identifiers with a number as prefix then I would cope that in a validation and offer a "quick fix" to change such a baseName input to a string literal.

Long story, short answer: **Are there any objection to this solution?**

**Comment by MathieuAA** [comment]
As long as an error is displayed when a user uses a digit as prefix, it's okay.

**Comment by bd82** [comment]
@MathieuAA 
> As long as an error is displayed when a user uses a digit as prefix, it's okay.

Yes there will be an error.

@colameo 
> I don't know your grammar, but maybe you can change it to a similar rule?

Its already identical to what you described.
